Automattic/woocommerce-payments
# UPE Block Checkout

Fixes #2162 

#### Changes proposed in this Pull Request

This PR adds the UPE to the block checkout experience. When UPE is enabled and the store is using the block checkout, UPE will now be loaded into the payment section. The block checkout makes use of the Stripe's React PaymentElement and so needed to update the `@stripe/react-stripe-js` package to the latest (1.4.1) this update also forced an update to the `@stripe/stripe-js` package to avoid any linting issues.

I kept the UPE block registration flow as independent as possible from the CC block registration flow in case there is ever a need to customize it without adding a lot of conditional checks. There is also the new UPE fields component to then use the PaymentElement. 

Some of the more notable details of the UPE fields component flow:
- The first was just that we need to create a payment intention before mounting the PaymentElement. I commented on the code here for any input on a better way to handle this just in case the approach I took isn't optimal. 
- Once mounted, there is an onChange event listener to check for the UPE form being completed.
- Once submitted, the onPaymentProcessing event will fire and check for the completed status and throw an error if not complete.
- In the classic checkout flow, we simulated a confirmPayment call to trigger the field errors. In this implementation I was unable to call the confirmPayment with an empty return_url as it caused it's own error. So the individual field errors will not be displayed unless the user has focused and blurred on them.
- Finally, the onCheckoutAfterProcessingWithSuccess is called after the server is updated. This part was tricky as we need to update the intent in case there was any price changes (different shipping method, coupon applied, etc). To update the intent required the order ID which was not available until this point. This leads to a slightly longer period between submitting the checkout and the final redirection from Stripe.

While working on this, we also got the feature from Stripe to pass in the billing details as needed and so that has been implemented here as part of the PR.









#### Testing instructions

- Start with a store with UPE enabled and Checkout blocks [enabled](https://docs.woocommerce.com/document/using-the-new-block-based-checkout/).
- Have 2 shipping methods, one flat rate (I used $10) and one free shipping.
- Add an item to the cart and select the flat rate shipping option from the cart.
- Confirm checkout loads properly and displays the UPE in the payments area.
- Try to submit without providing any payment details and confirm error is displayed.
- Provide only partial payment information, submit, and confirm error is displayed.
- Change the shipping method to the free shipping option, confirm the checkout total updates to reflect the change.
- Provide full payment information and submit
- Confirm that the checkout is successful 
- Confirm that the order was for the correct total (using free shipping)
- From the admin, view the transaction and confirm that customer details (name, email, address, etc) were correctly recorded for the transaction
- Run through the above steps using Euros and other payment methods (Giropay and Sofort)

Due to the update in Stripe JS packages, perform additional regression testing with: Classic CC checkout, Block CC checkout, Classic UPE checkout. Confirm all functionality appears unaffected. 





-------------------
